private_sharing/views.py

```diff
@@ -676,3 +686,123 @@ def get_object(self, queryset=None):
+        """
+        self.object = self.get_object()
+        if self.object.approved:
+            django_messages.error(
+                self.request,
+                (
+                    "Sorry, {0} has been approved and the project's datatypes cannot be changed "
+                    "without re-approval.".format(self.object.name)
+                ),
+            )
+
+            return HttpResponseRedirect(
+                reverse(
+                    "direct-sharing:detail-{0}".format(self.object.type),
+                    kwargs={"slug": self.object.slug},
+                )
+            )
+        return super().dispatch(*args, **kwargs)
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        fields = []
+        html_tab = SafeString("&emsp;&emsp;")
+
+        # Pre-populate form based on whether we saved the form in the session to add
+        # datatypes or, failing that, if there are datatypes already associated with
+        # the project
+        populate = self.request.session.pop("select_category_form", None)
+        if not populate:
+            populate = {}
+            for datatype in self.object.datatypes.all():
+                populate[datatype.html_safe_name] = ["on"]
+
+        for entry in DataType.objects.all().order_by("name"):
+            parents = entry.all_parents
+            if not entry.parent:
+                tab = ""  # We are not going to tab over the first level of the ontology
+            else:
+                tab = html_tab * len(parents)
+            if populate:
+                initial = populate.pop(entry.html_safe_name, False)
+            else:
+                initial = False
+            if initial == ["on"]:
+                initial = True
+            new_field = {
+                "label": entry.name,
+                "id": "id_{0}".format(entry.html_safe_name),
+                "initial": initial,
+                "name": entry.html_safe_name,
+                "description": entry.description,
+                "tab": tab,
+            }
+            if entry.parent:
+                for field in fields:
+                    if entry.parent.name == field["label"]:
+                        loc = fields.index(field)
+                        fields.insert(loc + 1, new_field)
+                        break
+            if new_field not in fields:
+                fields.append(new_field)
+
+        context.update({"fields_tree": fields})
+        return context
+
+    def form_valid(self, form):
+        """
+        Form is good, let's save this thing.
+        """
+        ret = super().form_valid(form)
+        if "add-datatype" in form.cleaned_data:
+            # User wants to add a new datatype.  Save the current state of the
+            # selection form.
+            form.cleaned_data.pop("add-datatype")
+            self.request.session["select_category_form"] = form.cleaned_data
```

Comment:
it strikes me as very unusual to be using session when filling in forms, I think I'd like to understand why it's being done this way (assuming you're confident this is the correct thing to do)

Reply:
I'm just using the session to temporarily save form state so that when we return to the form we can reload it.